File: JavaScript/binary_search.js


//function which returns the position of the element from an array
function binary_search(sortedArray, key) {
  //if the value is not present it simply returns -1
  let start = 0;
  let end = sortedArray.length - 1;

  while (start <= end) {
    let middle = Math.floor((start + end) / 2);

    if (sortedArray[middle] === key) {
      // found the key
      return middle;
    } else if (sortedArray[middle] < key) {
      // continue searching to the right
      start = middle + 1;
    } else {
      // search searching to the left
      end = middle - 1;
    }
  }
  // At each run of this while loop we reduce the size of the array to be searched by 2.

  return -1;
}

function main() {
  // We can only apply binary search if array is sorted so the input array must be sorted
  const arr = [2, 3, 14, 54, 63, 78];

  // Element to be searched in the array
  const key = 14;

  const pos = binary_search(arr, key);
  if (pos !== -1) {
    console.log("Element found in position", pos);
  } else {
    console.log("Element not found in array");
  }
  return 0;
}

main();

/*
Input: Sorted array of numbers and the number we are searching
Output: Returns position of the element if it's present if not -1

----------------------------------------
Sample Input:
[-5, 0, 6, 7, 8, 19]
6
Output:
Element found in position 3

Sample Input:
[-5, 0, 6, 7, 8, 19]
4
Output:
Element not found in array

----------------------------------------

Time Complexity:
Worst and Average case = O(logn)
Best case = O(1)

Space Complexity: O(1)
*/

File: Python/binarySearch.py
def binarySearch(x, u):
	l = 0
	h = len(x) - 1
	while l <= h:
		m = (h + l) // 2
		# Right half
		if x[m] < u:
			l = m + 1

		# Left half
		elif x[m] > u:
			h = m - 1

		# Mid
		else:
			return m
	# Not in list
	return False

x = list(map(int,input("Enter the list of numbers separated by space: ").split()))
u = int(input("Enter the number you want to search: "))

res = binarySearch(x, u)

if res != False:
	print("Number is present at:", str(res))
else:
	print("Number is not preset in list")


'''
Input:  x = 0 1 2 3 4 5 6 7 8 9 10
	u = 8
Output: Number is present at: 8

Average Time complexity: O(logN)
Space complexity: O(1)
'''
